### DISCUSSION TITLE
Progress working  sharepoint 2019 connector try CRUD queries

### DISCUSSION THREAD

**Discussion by vevedh** [Budibase/budibase#7481] [Contributing]
Hello,

I start integration of <a href="https://github.com/koltyakov/sp-jsom-node">https://github.com/koltyakov/sp-jsom-node</a>

I progress create a sharepoint 2013,2016,2019 connector to CRUD sharepoint Lists

<img width="325" alt="image" src="https://user-images.githubusercontent.com/1430389/187007794-5b5607fb-c44b-4ced-854f-1f1f10fd1a11.png">

<img width="377" alt="image" src="https://user-images.githubusercontent.com/1430389/187007802-2da3267f-1bc4-429a-a347-b2047784ca0d.png">

<img width="377" alt="image" src="https://user-images.githubusercontent.com/1430389/187007789-d6c6392c-9213-4a2e-8cd6-af595ece0b7a.png">

**Comment by poirazis** [comment]
That's nice @vevedh !

**Comment by vevedh** [comment]
Hello , 

When I call test Read to a Query :
<img width="488" alt="image" src="https://user-images.githubusercontent.com/1430389/187193736-8ad944a2-fe76-4a67-ae9d-903f383f8440.png">

I've this error : 
<img width="354" alt="image" src="https://user-images.githubusercontent.com/1430389/187193955-57a4e42d-7482-4373-8c52-381ea89ace0f.png">
<img width="307" alt="image" src="https://user-images.githubusercontent.com/1430389/187194010-f94c6d9c-cc67-438c-abc6-2bbe7eacca83.png">
<img width="489" alt="image" src="https://user-images.githubusercontent.com/1430389/187194066-2a9fc41e-3ab3-4d53-88c0-ca806c7b7230.png">

A part of my code 
 class SP2019Integration implements IntegrationBase {
    private readonly config: sp2019Config
    private client: JsomNode

    constructor(config: sp2019Config) {
      this.config = config
      const sp2019: JsomNode = new JsomNode({
        modules: ["taxonomy", "userprofiles"],
      })
      this.client = sp2019.init({
        siteUrl: this.config.siteUrl,

        authOptions: {
          username: this.config.username,
          password: this.config.password,
          domain: this.config.domain,
        },
      })
    }

    async read(query: { bucket: string }) {
      const response = await new Promise((resolve, reject) => {
        const ctx: SP.ClientContext = this.client.getContext()
        const oListsCollection = ctx.get_web().get_lists()
        ctx.load(oListsCollection, "Include(Title)")
        ctx
          .executeQueryPromise()
          .then(() => {
            resolve({ result: oListsCollection })
          })
          .catch(err => {
            throw new Error(`Sharepoint error: ${err}`)
          })
      })

      return response
    }
  }

**Comment by melohagan** [comment]
Hey @vevedh 

Thanks for having a go at an integration! 

If the server's crashing then you may be getting an uncaught exception. 

I think this could be the problem:
```js
.catch(err => {
        throw new Error(`Sharepoint error: ${err}`)
      })
```

I would try replacing it with: 
```js
.catch(err => {
  reject("Sharepoint error")
})
```

**Reply by vevedh** [reply]
Hello @melohagan 
With koa backend unsing the library sp-jsom-node crash.

I found the solution, I've create a docker container with a simple rest express  listen port 9090.
And I change my my script for this and It's perfect :
import {
    DatasourceFieldType,
    DatasourcePlus,
    Integration,
    IntegrationBase,
    QueryType,
  } from "@budibase/types"
  
  module SP2019Module {
    interface sp2019Config {
      siteUrl: string
      username: string
      password: string
      domain: string
    }
  
    const SCHEMA: Integration = {
      // Optional link to docs, which gets shown in the UI.
      docs: "https://github.com/koltyakov/sp-jsom-node",
      friendlyName: "SharePoint2019",
      type: "Non-relational",
      description:
        "SharePoint2019 Service to manage sharepoint 2019 on-premise server. ",
      datasource: {
        siteUrl: {
          type: "string",
          default: "http://svrsharepoint4.agglo.local",
          required: true,
        },
        username: {
          type: "string",
          default: "username",
          required: true,
        },
        password: {
          type: "password",
          default: "password",
          required: true,
        },
        domain: {
          type: "string",
          default: "domain",
          required: true,
        },
      },
      query: {
        command: {
          readable: true,
          displayName: "Sharepoint 2019 Get All Lists",
          type: QueryType.JSON,
        },
        create: {
          type: QueryType.FIELDS,
          fields: {
            key: {
              type: DatasourceFieldType.STRING,
              required: true,
            },
            value: {
              type: DatasourceFieldType.STRING,
              required: true,
            }
          },
        },
        read: {
          readable: true,
          type: QueryType.FIELDS,
          fields: {
            key: {
              type: DatasourceFieldType.STRING,
              required: true,
            },
          },
        },
        update: {
            readable: true,
            type: QueryType.FIELDS,
            fields: {
              key: {
                type: DatasourceFieldType.STRING,
                required: true,
              },
            },
          },
        delete: {
          type: QueryType.FIELDS,
          fields: {
            key: {
              type: DatasourceFieldType.STRING,
              required: true,
            },
          },
        }
      },
    }
  
    class SP2019Integration {
      private readonly config: sp2019Config
  
      constructor(config: sp2019Config) {
        this.config = config
      }
  
      request(ctx: any, request: any) {
        const { Headers } = require("@budibase/backend-core/constants")
        const {
          getTenantId,
          isTenantIdSet,
        } = require("@budibase/backend-core/tenancy")
        const env = require("../environment")
  
        if (!request.headers) {
          request.headers = {}
        }
        if (!ctx) {
          request.headers[Headers.API_KEY] = env.INTERNAL_API_KEY
          if (isTenantIdSet()) {
            request.headers[Headers.TENANT_ID] = getTenantId()
          }
        }
        if (request.body && Object.keys(request.body).length > 0) {
          request.headers["Content-Type"] = "application/json"
          request.body =
            typeof request.body === "object"
              ? JSON.stringify(request.body)
              : request.body
        } else {
          delete request.body
        }
        if (ctx && ctx.headers) {
          request.headers = ctx.headers
        }
        return request
      }
  
      async spContext(query: Function) {
        try {
          return await query()
        } catch (err) {
          throw new Error(`Redis error: ${err}`)
        } finally {
          console.log("end")
        }
      }
  
      async getSP2019() {
        try {
          const nodeFetch = require("node-fetch")
          const env = require("../environment")
          const result = await nodeFetch(
            //env.WORKER_URL+'/api/global/sp2019/lists', using sp-jsom-node  crash in worker
            // to solve run : docker run -p 9090:9090 --name=budi-sharepoint --hostname=localhost -it vevedh/budi-vvexpress
            "http://localhost:9090/lists",
            this.request(null, {
              method: "POST",
              body: {
                siteUrl: this.config.siteUrl,
                username: this.config.username,
                password: this.config.password,
                domain: this.config.domain,
              },
            })
          )
  
          const response = await result.json()
          console.log("TEST SHAREPOINT :", response)
          return response
        } catch (error) {
          console.log(`Sharepoint error: ${error}`)
          return { rows: error }
        }
      }
  
      async create(query: { key: string; value: string; ttl: number }) {
        return this.spContext(async () => {
          const response = null
          return response
        })
      }
  
      async read(query: { key: string }) {
        return this.spContext(async () => {
          const response = null
          return response
        })
      }

      async update(query: { key: string }) {
        return this.spContext(async () => {
          const response = null
          return response
        })
      }
  
      async delete(query: { key: string }) {
        return this.spContext(async () => {
          const response = null
          return response
        })
      }
  
      async command(query: { json: string }) {
        return this.spContext(async () => {
          const result = await this.getSP2019()
          console.log("Sharepoint list :", result)
          return result
        })
      }
    }
  
    module.exports = {
      schema: SCHEMA,
      integration: SP2019Integration,
    }
  }

Screens :  
<img width="368" alt="image" src="https://user-images.githubusercontent.com/1430389/188312457-38aca8b9-3bf2-4ef7-a4d0-67680b6883ee.png">

<img width="343" alt="image" src="https://user-images.githubusercontent.com/1430389/188312496-4e7fdaee-54d2-456e-beb2-2fd21746ac55.png">

<img width="342" alt="image" src="https://user-images.githubusercontent.com/1430389/188312523-d7a6bcbc-54e7-4672-9939-05e389e29e5f.png">

<img width="339" alt="image" src="https://user-images.githubusercontent.com/1430389/188312538-ba3421db-2cd7-45fd-8fb2-033edd4273b7.png">

<img width="339" alt="image" src="https://user-images.githubusercontent.com/1430389/188312558-5f76de0d-d226-4272-b89d-570edf99cb44.png">

<img width="467" alt="image" src="https://user-images.githubusercontent.com/1430389/188312590-276f187c-5019-4435-a5b2-cd940256e6d2.png">

**Comment by vevedh** [comment]
Hello, the team
Next working on CRUD sharepoint Lists Items.
How I can be a contributor , or developer member (french member) ?
